docs/proposals/push-based-metrics-collection.md

```diff
@@ -0,0 +1,77 @@
+
+![](../images/push-based-metrics-collection.png)
+Fig.1 Architecture of the new design
+
+## Goal
+1. **A new parameter in Python SDK function `tune`**: allow users to specify the method of collecting metrics(push-based/pull-based).
+2. **A code injection function in mutating webhook**: recognize the metrics output lines and replace them with push-based metrics collection code.
+3. The final metrics of worker pods should be **pushed to Katib DB directly** in the push mode of metrics collection.
+
+## API
+
+### New Parameter in Python SDK Function `tune`
+
+We decided to add `metrics_collection_mechanism` to `tune` function in Python SDK.
+
+```Python
+def tune(
+    self,
+    name: str,
+    objective: Callable,
+    parameters: Dict[str, Any],
+    base_image: str = constants.BASE_IMAGE_TENSORFLOW,
+    namespace: Optional[str] = None,
+    env_per_trial: Optional[Union[Dict[str, str], List[Union[client.V1EnvVar, client.V1EnvFromSource]]]] = None,
+    algorithm_name: str = "random",
+    algorithm_settings: Union[dict, List[models.V1beta1AlgorithmSetting], None] = None,
+    objective_metric_name: str = None,
+    additional_metric_names: List[str] = [],
+    objective_type: str = "maximize",
+    objective_goal: float = None,
+    max_trial_count: int = None,
+    parallel_trial_count: int = None,
+    max_failed_trial_count: int = None,
+    resources_per_trial: Union[dict, client.V1ResourceRequirements, None] = None,
+    retain_trials: bool = False,
+    packages_to_install: List[str] = None,
+    pip_index_url: str = "https://pypi.org/simple",
+    metrics_collection_mechanism: str = "pull", # The newly added parameter
+)
+```
+
+## Implementation
+
+### Add New Parameter in `tune`
+
+As is mentioned above, we decided to add `metrics_collection_mechanism` to the tune function in Python SDK. Also, we have some changes to be made:
+
+1. Disable injection: set `katib.kubeflow.org/metrics-collector-injection` to `disabled` when the push-based way of metrics collection is adopted so as to disable the injection of the metrics collection sidecar container.
+
+2. Configure the way of metrics collection: set the configuration `spec.metricsCollectionSpec.collector.kind`(specify the way of metrics collection) to `NoneCollector`.
+
+### Code Injection in Webhook
+
+We decided to implement a code replacing function in Experiment Mutating Webhook. When `spec.metricsCollectionSpec.collector.kind` is set to `NoneCollector`, the code replacing function will recognize the metrics output lines (e.g. print, log.Info, e.t.c.) and replace them with push-based metrics collection code which will be discussed in the next section. It’s a better decision compared with offering users a `katib_client.push`-like interface, for that users can’t use a yaml file to define this operation.
```

Comment:
I am not sure if we should replace user code here. I would suggest, we should implement it using `katib_client.report_metrics("lr": 0.03)` or `katib_client.log_metrics("lr": 0.01")`. That will allow user to explicitly report metrics during training code to the Katib DB.
Also, we should always pass `KATIB_TRIAL_NAME` env variable to the Katib Trials, so training function knows where to report metrics to the Katib DB.

Reply:
That's what I want to discuss with you in the upcoming weekly meeting.

Reply:
We can implement it with `katib_client.report_metrics` and call this function when we create Trial with Python SDK. However, users sometimes apply a yaml file instead of using Python API. If we want to make the push-based metrics collection compatible with this scenario, we should make some changes to the user code, for example, importing katib library and calling `katib_client.report_metrics`. Is it a better choice? WDYT👀@andreyvelich

Reply:
>However, users sometimes apply a yaml file instead of using Python API. If we want to make the push-based metrics collection compatible with this scenario, we should make some changes to the user code

That's right, users can build docker image and submit YAML.
In that case, they need to install `kubeflow-katib` Python SDK and use `katib_client.report_metrics()` API to report metrics from their training code.
I don't think we should do any changes to the user code, tbh.
What do you think @Electronic-Waste @tenzen-y @johnugeorge ?

Reply:
Yeah... Maybe my expressions are ambiguous. I mean users may need to change their code(importing kubeflow-katib and using `katib_client.report_metrics()`). It seems that the change they'll make is not bothering and difficult. I agree with you and I'll modify the KEP soon. @andreyvelich